website/orders/static/orders/js/admin-scanner.js

```diff
@@ -0,0 +1,179 @@
+                    callback_error.apply(args);
+                }
+                else {
+                    args.unshift(data.products);
+                    callback_ok.apply(this, args);
+                }
+            }}).fail(function() {
+                args.unshift("Error while getting product data.");
+                callback_error.apply(this, args);
+            });
+        }
+    )
+}
+
+function create_popup_button(product) {
+    let base = create_element("input", ["popup-button", "btn-ml", "btn-on"], product.name);
+    base.type = "button";
+    base.value = product.name;
+    base.addEventListener("click", add_product.bind(self, DATA_URL, CSRF_TOKEN, product_added_succeeded, product_added_failed, product));
+    return base;
+}
+
+function stop_scanner() {
+    Quagga.stop();
+    SCANNER.style.display = "none";
+    SCANNER_START_BUTTON.style.display = "";
+}
+
+function product_scanned(products) {
+    if (products.length > 0) {
+        stop_scanner();
+        scanned_codes = [];
+        POPUP_MODAL_BODY.innerHTML = "";
+        for (let i = 0; i < products.length; i++) {
+            let button = create_popup_button(products[i]);
+            POPUP_MODAL_BODY.append(button);
+        }
+        $(POPUP_MODAL).modal("show");
+    }
+}
+
+function product_error() {
+    console.log("No product data for this barcode.");
+}
+
+function scan_result(result) {
+    let code = result.codeResult.code;
+    if (!scanned_codes.includes(code)) {
+        scanned_codes.push(code);
+        get_product_details_from_barcode(DATA_URL, CSRF_TOKEN, product_scanned, product_error, code);
+    }
+}
+
+function add_product(data_url, csrf_token, callback_ok, callback_error, product /*, args */) {
+    $(POPUP_MODAL).modal("hide");
+    let args = Array.prototype.slice.call(arguments, 5);
+    jQuery(function($) {
+        let data = {
+            'csrfmiddlewaretoken': csrf_token,
+            'option': 'add',
+            'product': product.id
+        };
+        $.ajax({type: 'POST', url: data_url, data, dataType:'json', asynch: true, success:
+            function(data) {
+                if (data.error) {
+                    args.unshift(data.error);
+                    callback_error.apply(args);
+                }
+                else {
+                    args.unshift(data);
+                    callback_ok.apply(this, args);
+                }
+            }}).fail(function() {
+                args.unshift("Error while adding a product.");
+                callback_error.apply(this, args);
+            });
+        }
+    )
+}
+
+function product_added_succeeded() {
+    MESSAGE_CONTAINER.innerHTML = "";
+    MESSAGE_CONTAINER.append(create_element("p", ["alert", "alert-success"], "Product added successfully!"))
+}
+
+function product_added_failed(errormsg) {
+    MESSAGE_CONTAINER.innerHTML = "";
+    MESSAGE_CONTAINER.append(create_element("p", ["alert", "alert-danger"], errormsg))
+}
+
+start_scanner();
```

Comment:
I think it would be nicest to make the scanner page a simple form with a single field for the product code (that can be entered manually!), and with this scanner only filling that field. 
I think that would make the whole thing more user friendly, as you can also enter the code manually if you want.

The 'confirm the product' screen is undesirable I think, as it slows down the flow. What do you think? Maybe only give a 'undo' button?

Reply:
The problem is that the scanner is not really fault-proof. I'm not sure what the barcodes of the items in the canteen look like but if there are two that really look alike the scanner might pick something up that should not be there and might scan the wrong item. This is why I implemented the "select the product screen".

I can change it to this:

- The scanner automatically starts
- If the scanner picks up a correct barcode, it is added to a list of items at the bottom of the page.
- There are crosses next to the items that let the user delete items that are not wanted
- If a user clicks "Add to shift" at the bottom of the page, all items of the list get added.

Sound good?

Reply:
> The problem is that the scanner is not really fault-proof. I'm not sure what the barcodes of the items in the canteen look like but if there are two that really look alike the scanner might pick something up that should not be there and might scan the wrong item. This is why I implemented the "select the product screen".
> 
EAN codes have a checksum so this shouldn't be a problem I think.

Benefit of such a field would be that we can also easily support 'real' manual scanners


> I can change it to this:
> 
> * The scanner automatically starts
> * If the scanner picks up a correct barcode, it is added to a list of items at the bottom of the page.
> * There are crosses next to the items that let the user delete items that are not wanted
> * If a user clicks "Add to shift" at the bottom of the page, all items of the list get added.
> 
> Sound good?

Sounds more complex than what we currently have, so lets not do this then.

Reply:
> > The problem is that the scanner is not really fault-proof. I'm not sure what the barcodes of the items in the canteen look like but if there are two that really look alike the scanner might pick something up that should not be there and might scan the wrong item. This is why I implemented the "select the product screen".
> 
> EAN codes have a checksum so this shouldn't be a problem I think.

Ahh, nice, then I'll implement a checksum checker in javascript (which will be a lot of fun, not).

> Benefit of such a field would be that we can also easily support 'real' manual scanners
> 
> > I can change it to this:
> > 
> > * The scanner automatically starts
> > * If the scanner picks up a correct barcode, it is added to a list of items at the bottom of the page.
> > * There are crosses next to the items that let the user delete items that are not wanted
> > * If a user clicks "Add to shift" at the bottom of the page, all items of the list get added.
> > 
> > Sound good?
> 
> Sounds more complex than what we currently have, so lets not do this then.

Okay, another idea:

- The scanner will NOT open a new page but rather a popup.
- The scanner will check the EAN codes for the checksum, if a checksum is correct it will be send to the server to look up the associated item.
- If the barcode exists in the database, we will add it immediately, close the scanner and put a notification on top of the page saying the product was added successfully.
- I will also put a search field on top of the scanner where an admin can search for either the barcode or the product name and add it manually.

Reply:
Sounds good, maybe even faster to do the matching on-device in javascript already. 
Maybe use a javascript alert()?

Reply:
By the way, I assume that our library does an automatic checksum check already

Reply:
> Sounds good, maybe even faster to do the matching on-device in javascript already.
> Maybe use a javascript alert()?

Then we need to ship a list of all objects with barcode to the web browser which is not an issue with a couple of products but if we have a lot it is not the best way of doing this.

Django is fast enough to respond almost instantly so it's not a problem to check serverside.

Reply:
> By the way, I assume that our library does an automatic checksum check already

I thought that too but I saw some weird codes ending up in the logs, I'll make sure to check this.

Reply:
> > By the way, I assume that our library does an automatic checksum check already
> 
> I thought that too but I saw some weird codes ending up in the logs, I'll make sure to check this.

You were correct, it checks the codes.

Reply:
This is also changed in the latest commit, could you please check again if you agree with the changes?